### ISSUE TITLE
Empty State secondary action buttons do not align properly when labels differ in length

### ISSUE THREAD

**Issue by beanh66** [patternfly/patternfly-3#907]
In OpenShift we came across an issue where adding three secondary action buttons underneath the primary action (as the PF Empty State design suggests) created an issue with the alignment. The resolution used in OpenShift includes flexbox and only fixes the case where exactly 3 secondary actions are added. A more robust solution should be implemented in PF core to account for various cases. 

For reference: [OpenShift GitHub issue 2644](https://github.com/openshift/origin-web-console/pull/2644)

**Comment by beanh66** [comment]
FYI @jennyhaines @serenamarie125 @dtaylor113

**Comment by jennyhaines** [comment]
@beanh66 - I'll think about this more and mock up some ideas!

**Comment by andresgalante** [comment]
@beanh66 This is a good idea, when blank slate was added we supported IE9 which left flexbox out of scope. Now that we don't we can change it.

Do you think you can find time to PR this? if not we can do it cc/ @LHinson

**Comment by beanh66** [comment]
@andresgalante I'm not sure, @rhamilto did the work for the OpenShift case, but at the moment it only covers the case where exactly 3 secondary actions exist. I think we want to be sure this covers the case where 0, 1, 2, or 3 secondary actions are used.

**Comment by jennyhaines** [comment]
@beanh66 @andresgalante -

I've mocked up the cases when 1, 2, 3, or 4-5 secondary actions are presented within an empty state! (When there are 0 secondary actions, just the primary action button will appear, so no need to mock up!) Thoughts? I thin this is pretty straight forward... I tried a lot of visual options out, but came back to this!

**1 Secondary Action:**
![optiona](https://user-images.githubusercontent.com/24898297/36318877-2327020c-130f-11e8-97c3-e108c2cec0e4.png)


**2 Secondary Actions:**
![2 secondary](https://user-images.githubusercontent.com/24898297/36318796-ecfb9120-130e-11e8-8da7-da85ee818758.png)


**3 Secondary Actions (would stay the same as before):**
![3 secondary](https://user-images.githubusercontent.com/24898297/36318805-f7d68136-130e-11e8-92ed-a78335cdc4c6.png)


**4-5 Secondary Actions (Fringe case. 5 is absolute max we should recommend):**
![4-5 secondary](https://user-images.githubusercontent.com/24898297/36318655-7b47f9ba-130e-11e8-9286-c55c9bf8b4c6.png)

This is how the links would look on mobile: 
***I think they should have 13px of clickable space around the text of each link. This would bring us up to 35px of vertical space for each link, which is over the 34px standard.
![4-5 secondary mobile](https://user-images.githubusercontent.com/24898297/36318709-ac7233de-130e-11e8-97f1-1a389d47d67e.png)

**Comment by beanh66** [comment]
Thanks @jennyhaines! As we discussed, I think these look great.

**Comment by beanh66** [comment]
@andresgalante Any update on this? @amysueg FYI this is the issue we discussed that may be helpful for your product as well :)

**Comment by amysueg** [comment]
@beanh66  - Will keep in mind if we ever have 3 or more actions in the empty state. at the moment I believe our two actions align as specified by patternfly.

**Comment by LHinson** [comment]
I've added this story to the PF backlog. It's in our current sprint and @dabeng is going to address it. Here is the link to the story if interested: https://patternfly.atlassian.net/browse/PTNFLY-2801

**Comment by dabeng** [comment]
Hi @jennyhaines . I created a **[demo page](https://rawgit.com/dabeng/patternfly/empty-state-dist/dist/tests/blank-slate.html)** of blank slate according to your design pictures. @andresgalante @LHinson please review the demo whenever you get the chance.

**Comment by beanh66** [comment]
@dabeng Thanks that looks great!

**Comment by dabeng** [comment]
@beanh66 It's my pleasure that this fix could be helpful for you 😊

**Comment by amarie401** [comment]
Looks awesome @dabeng !

**Comment by rhamilto** [comment]
A consideration:  button labels with a length longer than the specified min-width end up wider than buttons without.  Is this intentional?

![screen shot 2018-02-28 at 10 13 25 am](https://user-images.githubusercontent.com/895728/36795119-2a1d1b72-1c70-11e8-90fa-8f9e4cac979b.PNG)

My recommendation is to build and test using real world data, not mock data that may or may not illustrate such things.

**Comment by serenamarie125** [comment]
@beanh66 can you also review the demo page from @dabeng ?

**Comment by serenamarie125** [comment]
@dabeng looks good!
@rhamilto brings up a good point.  Can you update the storybook so we can see how it looks/behaves responsively when there is a long button label?

**Comment by beanh66** [comment]
@rhamilto Agreed. @dabeng button widths should be consistent regardless of the button label lengths. Can you make sure that is the case here with this implementation? @jennyhaines is there a certain button label max length we should note here as well? At some point we would need to cut it off and just show... with the full name on hover (not that buttons should ever be that long :)

**Comment by jennyhaines** [comment]
@dabeng - I think it looks great! The link might want to say "in the documentation" instead of "on the documentation", but that's nit-picky, I know! To address what @rhamilto noticed about the longer button length, I agree that the button widths should be consistent. 

To keep the button lengths from betting out of hand, we might want to add a suggested max character count for our button pattern in PatternFly. To account for translations in different [longer] languages, we may want to suggest buttons be no longer than 32 characters long. To visualize this, Robb's example of the long button that says "Secondary Action With a Longer Label" is 36 characters long. I've come to this "32 character" conclusion from researching [button length trends](https://explore.reallygoodemails.com/everything-you-wanted-to-know-about-email-cta-buttons-98807ab98806) and scouring the internet for answers. 
[1](https://ux.stackexchange.com/questions/55250/maximum-length-of-a-ui-element-string) - 32 character is the "magic max number" for CTAs
[2](https://sproutsocial.com/insights/social-media-character-counter/) - Facebook set a 25 character max for lead generation ad CTAs
[3](http://facebookmarketingpartners.com/wp-content/uploads/2016/03/Canvas-Creative-Specs-1-sheet_Jan2016.pdf) - Facebook set a 30 character max for their canvas ad CTAs

**Comment by beanh66** [comment]
Awesome, thanks for the input @jennyhaines 👍

**Comment by dabeng** [comment]
Hi @jennyhaines , I have updated the [demo page](https://rawgit.com/dabeng/patternfly/empty-state-dist/dist/tests/blank-slate.html) according to your idea. I choose to use unit ["ch"](http://meyerweb.com/eric/thoughts/2012/05/15/defining-ch/) to set button width.

The following line let secondary button hold the 32 "0" characters properly.
```css
@blank-slate-pf-secondary-btn-width: (calc(~"32ch + "(@padding-base-horizontal + @border-width-base) * 2));
```

Additionally, I have a concern. Due to "ch" is a relative unit, I have to construct the complex calc() expression in order to keep the width of primary button equal to secondary button width.

<img width="545" alt="screen shot 2018-03-01 at 5 35 28 pm" src="https://user-images.githubusercontent.com/5689075/36837886-874281bc-1d78-11e8-8f9f-bcee0c2da82d.png">

Here's the related less source code:
```css
@blank-slate-pf-primary-btn-width:   (calc(~"32ch * .8571 + "(ceil((@padding-large-horizontal + @border-width-base) * 2 * .6363))));
```
```js
0.8571  =  secondary button "font-size" / primary button "font-size"
0.6363 =  (secondary button "padding-left" + "border-width") / (primary button "padding-left" + "border-width")
```

Perhaps, it's more simple and easy-to-maintain to directly use fixed value - 180px to set action button width. 

Or we can provide users with two widths - 180px for common buttons with short label, and 240px for buttons with long label.

**Comment by dabeng** [comment]
0.8571  =  secondary button font-size / primary button font-size
0.6363 = (primary button padding-left + border-width) /  (secondary button padding-left + border-width)

**Comment by jennyhaines** [comment]
@dabeng - This is why I am not a dev... darn MATH! :-) Thank you for doing that! I do understand the need to make sure that the code is easy to maintain, so I think you are right in using set pixel widths. (180 for short buttons, and 240 for longer buttons).

My only worry is around differing language translation lengths. Something that is, say, 20 characters in English might be 30 characters in German. Is it possible for the code to do a check on how many characters are in each of the empty state buttons, and if one is more than "x" characters long, have the buttons in the empty state display at the larger button width? I'm just trying to think of a solution that accommodates different translation lengths. (FYI @andresgalante @dtaylor113 )

**Comment by dabeng** [comment]
@andresgalante  @jennyhaines @beanh66 @rhamilto I have created demos on codepen.io to show different solutions. They are easy to be tested and amended, so feel free to customize the code based on your ideas. 

- [Patternfly Blank State Action Buttons with ch unit](https://codepen.io/dabeng/pen/zRbwxJ)

- [Patternfly Blank State Action Buttons with px unit](https://codepen.io/dabeng/pen/NyJjdv)

Besides, I also added some labels composed of german characters  with the help google translate.

FYI, on condition with same language and same character encoding, different character has different width. So the width of 32 characters is not fixed value. My worry is that how to set uniform width attribute for two buttons which include 32 characters of different language.

**Comment by serenamarie125** [comment]
@dabeng thanks for this extra work!  

Regarding the actual length to use/algorithm, I would defer to suggestions from @andresgalante or @rhamilto on this

**Comment by beanh66** [comment]
@dabeng Thanks, this looks great 👍 Just posting a screenshot of an example I tried with your demo so others can see. 
<img width="807" alt="screen shot 2018-03-02 at 11 02 37 am" src="https://user-images.githubusercontent.com/18057044/36908492-5c5ea520-1e09-11e8-97b8-b7b31a8b135f.png">

**Comment by andresgalante** [comment]
Hi @beanh66 I have a design question: if the label is a single short word then then btn looks too wide, and if it's a large string, it get ellipse, which might be really bad for a call to action since you wouldn't be able to read it! I mean, it's a call to action it might something like: "Click on this btn to immediately **delete everything you've ever done**" and the last part can be cut off 😄.

I don't know how to solve the silly looking small label btn while maintaining the desirable effect, but my recommendation is to keep the `min-width` instead of `with` and let them be different if the label is larger.

@jennyhaines are you sure you want 13px? Although there is no rule, as far as I know we try to maintain spacer multiple of 5 for PF3.

@dabeng as you mentioned the `ch` is not a character count but relative to the width of the "0". It doesn't ensure 32 characters since for example an "m" is wider. I'd probably remove the calc and make that simpler but since this is a purely code related discussion I'd say let's review it on a PR.

**Comment by beanh66** [comment]
@andresgalante I see your point, but I don't know that we should account for ANY length button label. Realistically the names should never be that long IMO. I'd prefer the buttons be the same length but it seems to be more of a visual design question. I think in the rare case where labels are very long, consumers can take advantage of the link option rather than buttons. @jennyhaines provided these designs because it didn't look great with different length buttons in OpenShift.

**Comment by LHinson** [comment]
After looking through this issues, I see that @rhamilto says in the [OpenShift GitHub issue 2644](https://github.com/openshift/origin-web-console/pull/2644)

> I can [send a PR to PatternFly], but understand this fix is specific to three secondary action buttons. I think the PatternFly solution needs to be more robust in its approach to a variable number of buttons, and I'm not sure how to go about doing that. It seems like this should become an upstream issue to be solved by the community at large given the complexity of the problem and the issues with the existing empty state solution in PatternFly.

For the reason, I think this would mean that we keep the btns as they are in PatternFly today however, maybe we need developer notes to help guide upstream on how they can resolve this issue? 

WDYT @matthewcarleton @andresgalante?

**Comment by matthewcarleton** [comment]
@LHinson @andresgalante I agree. It seems this is not something that can be completely solved in core. It's a great opportunity to make use of patternfly.org dev notes to advise best practice.

**Comment by dabeng** [comment]
**Development notes of secondary action buttons of empty state pattern**

1 If width of secondary button is short (e.g. less than 300px), you can set the same width for all action buttons according to the widest button as shown as below:

![screenshot from 2018-07-05 20-22-38](https://user-images.githubusercontent.com/5689075/42323206-e7b83eb6-8091-11e8-9b7d-ecec3d79a73f.png)

2 If you have to add multiple action elements into secondary section(e.g. more than 5) or you have a secondary action element which includes too many words, we recommend you create them by the form of links.
![screenshot from 2018-07-05 20-37-27](https://user-images.githubusercontent.com/5689075/42323618-57841cc8-8093-11e8-8e0f-bd24f17adbbf.png)

Here's the **[Demo page](https://rawgit.com/dabeng/patternfly/empty-state-dist/dist/tests/blank-slate.html)**.

**Comment by dabeng** [comment]
@matthewcarleton @beanh66 @LHinson @jennyhaines need your new thoughts about above dev notes :blush:

**Comment by matthewcarleton** [comment]
looks good @dabeng. I'd suggest stating `min-width` instead of width. Also state is spelled incorrectly as "slate" :)

**Comment by dabeng** [comment]
Hi @matthewcarleton  I just followed naming convention used in [test page](https://rawgit.com/patternfly/patternfly/master-dist/dist/tests/blank-slate.html)

**Comment by matthewcarleton** [comment]
@dabeng ha, my bad :)

**Comment by mcarrano** [comment]
@dabeng I like the second option with links instead of buttons for secondary actions. But this is not specified in the current design. What do others think. Is this a variant that we should add? @beanh66 @jennyhaines interested in your thoughts.

**Comment by jennyhaines** [comment]
@mcarrano @dabeng - 
I can see the use of links as a secondary option for instances when those secondary actions are scarcely used.

**Comment by dabeng** [comment]
Hi @mcarrano The second option is derived from @jennyhaines [recommendation](https://github.com/patternfly/patternfly/issues/907#issuecomment-366291029)

> 4-5 Secondary Actions (Fringe case. 5 is absolute max we should recommend)

**Comment by mcarrano** [comment]
Sounds good to me. What are the next steps here @dabeng ? Are we ready to send a PR?

I will also open a design issue so the documentation can be updated.

**Comment by LHinson** [comment]
@dabeng @mcarrano in order to close out this particular issue, I'd like to add the dev notes for the original design, first example. This would be a PR against patternfly-org to add dev notes. If you would like to proceed with a new/alt design, let's proceed with that over in patternfly-design and follow up with implementation. However, I'd like to avoid scope creep for this particular issue. 

@dabeng any questions/concerns about this approach?

**Comment by mcarrano** [comment]
Makes sense @LHinson . I have opened a new design issue here that can live in our backlog to cover alternate secondary button options. https://github.com/patternfly/patternfly-design/issues/696

**Comment by dabeng** [comment]
LGTM +1

**Comment by LHinson** [comment]
@dabeng in order to close out this issue, dev notes are needed for the original design (your first example). This would be a PR against patternfly-org. I don't see the notes up on patternfly.org. What's the update on this?

**Comment by dabeng** [comment]
Hi @Dobflem I planned to send PR to patternfly-core repo to add new examples of empty state on testing page and then send the associated PR to patternfly-org repo

**Comment by LHinson** [comment]
@dabeng ok sounds good.

**Comment by jeff-phillips-18** [comment]
Fixed by #1117
